client/packages/programs/src/RnRForms/DetailView/RnRFormLine.tsx

```diff
@@ -0,0 +1,268 @@
+import React, { useState } from 'react';
+import {
+  BasicTextInput,
+  Checkbox,
+  DatePicker,
+  Formatter,
+  NumericTextInput,
+  useBufferState,
+  useNotification,
+  useTheme,
+  VenCategoryType,
+} from '@openmsupply-client/common';
+import { RnRFormLineFragment } from '../../api/operations.generated';
+
+export const RnRFormLine = ({
+  line,
+  saveLine,
+  periodLength,
+  disabled,
+}: {
+  line: RnRFormLineFragment;
+  periodLength: number;
+  saveLine: (line: RnRFormLineFragment) => Promise<void>;
+  disabled: boolean;
+}) => {
+  const theme = useTheme();
+  const { error } = useNotification();
+
+  const [patch, setPatch] = useState<Partial<RnRFormLineFragment>>({});
+  const draft = { ...line, ...patch };
+
+  const updateDraft = (update: Partial<RnRFormLineFragment>) => {
+    const newPatch = {
+      ...patch,
+      confirmed: false,
+      ...update,
+    };
+
+    const {
+      initialBalance,
+      quantityConsumed,
+      quantityReceived,
+      adjustments,
+      stockOutDuration,
+      previousAverageMonthlyConsumption,
+    } = { ...draft, ...newPatch };
+
+    const finalBalance =
+      initialBalance + quantityReceived - quantityConsumed + adjustments;
+
+    const stockAvailableDays = periodLength - stockOutDuration;
+
+    const adjustedQuantityConsumed = stockAvailableDays
+      ? quantityConsumed * (periodLength / stockAvailableDays)
+      : quantityConsumed;
+
+    // This calculation might be a plugin in future!
+    const averageMonthlyConsumption =
+      // Average of the last 3 months (including the current month)
+      // TODO: what if don't want to consider previous months?
+      (2 * previousAverageMonthlyConsumption + adjustedQuantityConsumed) / 3;
+
+    const maximumQuantity = averageMonthlyConsumption * 2;
+
+    const neededQuantity = maximumQuantity - finalBalance;
+    const requestedQuantity = neededQuantity > 0 ? neededQuantity : 0;
+
+    setPatch({
+      ...newPatch,
+      finalBalance,
+      adjustedQuantityConsumed,
+      averageMonthlyConsumption,
+      maximumQuantity,
+      requestedQuantity,
+    });
+  };
+
+  const venCategory =
+    draft.item.venCategory === VenCategoryType.NotAssigned
+      ? ''
+      : draft.item.venCategory;
+
+  const textColor =
+    disabled || draft.confirmed
+      ? theme.palette.text.disabled
+      : theme.palette.text.primary;
+
+  const readOnlyColumn = {
+    backgroundColor: theme.palette.background.drawer,
+    padding: '5px',
```

Comment:
:nit given that this is used in a `style` prop, it can be just numeric i.e. `padding: 5`